File: equibets/results.py
"""Result records, consolidation, and transparent finishing-score prediction."""

from __future__ import annotations

import json
import re
from collections import Counter, defaultdict
from dataclasses import dataclass
from datetime import date, datetime
from pathlib import Path
from statistics import mean, pstdev


USER_ENTERED_PRIORITY = 100


@dataclass(frozen=True)
class EventingResult:
    """One eventing result for a horse and rider combination."""

    source_id: str
    source_record_id: str
    source_priority: int
    rider_name: str
    horse_name: str
    event_name: str
    event_date: date
    level: str
    country: str
    dressage_score: float
    show_jumping_penalties: float
    cross_country_jump_penalties: float
    cross_country_time_penalties: float
    collected_at: datetime
    is_user_entered: bool = False
    rider_fei_id: str = ""
    horse_fei_id: str = ""
    combination_id: str = ""
    placing: str = ""
    show_jumping_time_penalties: float = 0.0
    total_score: float | None = None
    status: str = "completed"
    mer_status: str = ""
    event_url: str = ""
    source_url: str = ""
    venue: str = ""

    @property
    def finishing_score(self) -> float:
        """Lower scores are better in eventing."""

        if self.total_score is not None:
            return round(self.total_score, 1)
        return round(
            self.dressage_score
            + self.show_jumping_penalties
            + self.show_jumping_time_penalties
            + self.cross_country_jump_penalties
            + self.cross_country_time_penalties,
            1,
        )

    @property
    def is_completed(self) -> bool:
        return _completion_status(self.status, self.placing) == "completed"

    @property
    def combination_key(self) -> str:
        return f"{_slug(self.rider_name)}::{_slug(self.horse_name)}"

    @property
    def result_key(self) -> tuple[str, str, date, str]:
        return (
            self.combination_key,
            _slug(self.event_name),
            self.event_date,
            _slug(self.level),
        )

    @classmethod
    def from_mapping(cls, values: dict[str, object]) -> "EventingResult":
        is_user_entered = _optional_bool(values, "is_user_entered", False)
        return cls(
            source_id=_required_str(values, "source_id"),
            source_record_id=_required_str(values, "source_record_id"),
            source_priority=_optional_int(
                values,
                "source_priority",
                USER_ENTERED_PRIORITY if is_user_entered else 50,
            ),
            rider_name=_required_str(values, "rider_name"),
            horse_name=_required_str(values, "horse_name"),
            event_name=_required_str(values, "event_name"),
            event_date=date.fromisoformat(_required_str(values, "event_date")),
            level=_required_str(values, "level"),
            country=_required_str(values, "country"),
            dressage_score=_required_number(values, "dressage_score"),
            show_jumping_penalties=_required_number(values, "show_jumping_penalties"),
            cross_country_jump_penalties=_required_number(
                values,
                "cross_country_jump_penalties",
            ),
            cross_country_time_penalties=_required_number(
                values,
                "cross_country_time_penalties",
            ),
            collected_at=datetime.fromisoformat(_required_str(values, "collected_at")),
            is_user_entered=is_user_entered,
            rider_fei_id=_optional_str(values, "rider_fei_id"),
            horse_fei_id=_optional_str(values, "horse_fei_id"),
            combination_id=_optional_str(values, "combination_id"),
            placing=_optional_str(values, "placing"),
            show_jumping_time_penalties=_optional_number(values, "show_jumping_time_penalties", 0.0),
            total_score=_nullable_number(values, "total_score"),
            status=_optional_str(values, "status", "completed"),
            mer_status=_optional_str(values, "mer_status"),
            event_url=_optional_str(values, "event_url"),
            source_url=_optional_str(values, "source_url"),
            venue=_optional_str(values, "venue"),
        )


@dataclass(frozen=True)
class CombinationPrediction:
    """Likely upcoming finishing score for a combination."""

    rider_name: str
    horse_name: str
    likely_finishing_score: float
    recent_result_count: int
    best_recent_score: float
    worst_recent_score: float
    source_ids: tuple[str, ...]
    confidence: str
    predicted_low_score: float = 0.0
    predicted_high_score: float = 0.0


@dataclass(frozen=True)
class PredictionEvidence:
    """Transparent inputs behind a score-range prediction."""

    rider_name: str
    horse_name: str
    target_level: str
    predicted_dressage_score: float
    predicted_xc_jump_penalties: float
    predicted_xc_time_penalties: float
    predicted_show_jumping_penalties: float
    predicted_show_jumping_time_penalties: float
    predicted_final_score_low: float
    predicted_final_score_high: float
    average_dressage_current_level: float | None
    average_dressage_one_level_below: float | None
    average_final_score: float | None
    xc_clear_jumping_rate: float
    xc_time_penalty_average: float
    sj_rail_average: float
    completion_rate: float
    elimination_retirement_rate: float
    best_score: float | None
    worst_score: float | None
    recent_3_run_average: float | None
    recent_5_run_average: float | None
    trend_direction: str
    level_reliability: str
    country_pattern: str
    venue_pattern: str
    same_level_result_count: int
    horse_result_count: int
    combination_result_count: int


def load_results(path: Path | str) -> list[EventingResult]:
    """Load eventing results from JSON."""

    with Path(path).open(encoding="utf-8") as results_file:
        payload = json.load(results_file)

    return [EventingResult.from_mapping(item) for item in payload["results"]]


def consolidate_results(results: list[EventingResult]) -> list[EventingResult]:
    """Deduplicate results, keeping the highest-priority source for each start."""

    selected: dict[tuple[str, str, date, str], EventingResult] = {}
    for result in results:
        existing = selected.get(result.result_key)
        if existing is None or _is_better_result(result, existing):
            selected[result.result_key] = result

    return sorted(
        selected.values(),
        key=lambda result: (result.event_date, result.rider_name, result.horse_name),
    )


def predict_finishing_score(
    results: list[EventingResult],
    rider_name: str,
    horse_name: str,
    *,
    recent_result_limit: int = 5,
) -> CombinationPrediction:
    """Estimate likely finishing score from recent consolidated starts."""

    combination_key = f"{_slug(rider_name)}::{_slug(horse_name)}"
    recent_results = [
        result
        for result in consolidate_results(results)
        if result.combination_key == combination_key
    ]
    recent_results = sorted(
        recent_results,
        key=lambda result: result.event_date,
        reverse=True,
    )[:recent_result_limit]

    if not recent_results:
        raise ValueError("No results found for combination")

    weighted_score_total = 0.0
    weight_total = 0
    for weight, result in zip(range(len(recent_results), 0, -1), recent_results):
        weighted_score_total += result.finishing_score * weight
        weight_total += weight

    scores = [result.finishing_score for result in recent_results]
    evidence = build_prediction_evidence(
        results,
        recent_results[0].rider_name,
        recent_results[0].horse_name,
        target_level=recent_results[0].level,
    )
    return CombinationPrediction(
        rider_name=recent_results[0].rider_name,
        horse_name=recent_results[0].horse_name,
        likely_finishing_score=round(weighted_score_total / weight_total, 1),
        recent_result_count=len(recent_results),
        best_recent_score=min(scores),
        worst_recent_score=max(scores),
        source_ids=tuple(sorted({result.source_id for result in recent_results})),
        confidence=_confidence(len(recent_results)),
        predicted_low_score=evidence.predicted_final_score_low,
        predicted_high_score=evidence.predicted_final_score_high,
    )


def build_prediction_evidence(
    results: list[EventingResult],
    rider_name: str,
    horse_name: str,
    *,
    target_level: str | None = None,
) -> PredictionEvidence:
    """Calculate prediction inputs for one horse/rider combination."""

    consolidated = sorted(consolidate_results(results), key=lambda result: result.event_date, reverse=True)
    horse_key = _slug(horse_name)
    combination_key = f"{_slug(rider_name)}::{horse_key}"
    horse_results = [result for result in consolidated if _slug(result.horse_name) == horse_key]
    combination_results = [result for result in horse_results if result.combination_key == combination_key]
    if not combination_results and not horse_results:
        raise ValueError("No results found for horse or combination")

    evidence_results = combination_results or horse_results
    selected_level = target_level or evidence_results[0].level
    completed = [result for result in evidence_results if result.is_completed]
    same_level = [result for result in completed if _same_level(result.level, selected_level)]
    one_below = [result for result in completed if _level_rank(result.level) == _level_rank(selected_level) - 1]
    score_values = [result.finishing_score for result in completed]

    predicted_dressage = _weighted_phase_average(evidence_results, selected_level, "dressage_score")
    predicted_xc_jump = _weighted_phase_average(evidence_results, selected_level, "cross_country_jump_penalties")
    predicted_xc_time = _weighted_phase_average(evidence_results, selected_level, "cross_country_time_penalties")
    predicted_sj_jump = _weighted_phase_average(evidence_results, selected_level, "show_jumping_penalties")
    predicted_sj_time = _weighted_phase_average(evidence_results, selected_level, "show_jumping_time_penalties")
    predicted_total = predicted_dressage + predicted_xc_jump + predicted_xc_time + predicted_sj_jump + predicted_sj_time

    completion_rate = _rate(len(completed), len(evidence_results))
    el_ret_rate = _rate(
        sum(1 for result in evidence_results if _completion_status(result.status, result.placing) in {"eliminated", "retired", "withdrawn"}),
        len(evidence_results),
    )
    uncertainty = max(2.5, (pstdev(score_values) if len(score_values) > 1 else 3.0) + el_ret_rate * 8.0)

    return PredictionEvidence(
        rider_name=evidence_results[0].rider_name,
        horse_name=evidence_results[0].horse_name,
        target_level=selected_level,
        predicted_dressage_score=round(predicted_dressage, 1),
        predicted_xc_jump_penalties=round(predicted_xc_jump, 1),
        predicted_xc_time_penalties=round(predicted_xc_time, 1),
        predicted_show_jumping_penalties=round(predicted_sj_jump, 1),
        predicted_show_jumping_time_penalties=round(predicted_sj_time, 1),
        predicted_final_score_low=round(max(0.0, predicted_total - uncertainty), 1),
        predicted_final_score_high=round(predicted_total + uncertainty, 1),
        average_dressage_current_level=_average([result.dressage_score for result in same_level]),
        average_dressage_one_level_below=_average([result.dressage_score for result in one_below]),
        average_final_score=_average(score_values),
        xc_clear_jumping_rate=_rate(sum(1 for result in completed if result.cross_country_jump_penalties == 0), len(completed)),
        xc_time_penalty_average=_average([result.cross_country_time_penalties for result in completed]) or 0.0,
        sj_rail_average=round((_average([result.show_jumping_penalties for result in completed]) or 0.0) / 4.0, 2),
        completion_rate=completion_rate,
        elimination_retirement_rate=el_ret_rate,
        best_score=min(score_values) if score_values else None,
        worst_score=max(score_values) if score_values else None,
        recent_3_run_average=_average(score_values[:3]),
        recent_5_run_average=_average(score_values[:5]),
        trend_direction=_trend(score_values),
        level_reliability=_level_reliability(same_level, one_below, completion_rate, el_ret_rate),
        country_pattern=_pattern(evidence_results, "country"),
        venue_pattern=_pattern(evidence_results, "venue"),
        same_level_result_count=len(same_level),
        horse_result_count=len(horse_results),
        combination_result_count=len(combination_results),
    )


def _is_better_result(candidate: EventingResult, existing: EventingResult) -> bool:
    if candidate.source_priority != existing.source_priority:
        return candidate.source_priority < existing.source_priority
    if candidate.is_user_entered != existing.is_user_entered:
        return not candidate.is_user_entered
    return candidate.collected_at > existing.collected_at


def _confidence(result_count: int) -> str:
    if result_count >= 5:
        return "high"
    if result_count >= 3:
        return "medium"
    return "low"


def _weighted_phase_average(results: list[EventingResult], target_level: str, field_name: str) -> float:
    values: list[tuple[float, float]] = []
    ordered = sorted(results, key=lambda result: result.event_date, reverse=True)
    for index, result in enumerate(ordered):
        if not result.is_completed:
            continue
        value = getattr(result, field_name)
        recency_weight = max(1.0, len(ordered) - index)
        level_weight = 1.5 if _same_level(result.level, target_level) else 0.85
        if _level_rank(result.level) > _level_rank(target_level):
            level_weight = 1.2
        values.append((float(value), recency_weight * level_weight))
    if not values:
        return 0.0
    return sum(value * weight for value, weight in values) / sum(weight for _, weight in values)


def _average(values: list[float]) -> float | None:
    if not values:
        return None
    return round(mean(values), 1)


def _rate(count: int, total: int) -> float:
    if total == 0:
        return 0.0
    return round(count / total, 2)


def _trend(scores_descending: list[float]) -> str:
    if len(scores_descending) < 4:
        return "flat"
    newer = mean(scores_descending[: max(2, len(scores_descending) // 2)])
    older = mean(scores_descending[max(2, len(scores_descending) // 2) :])
    if newer <= older - 1.0:
        return "improving"
    if newer >= older + 1.0:
        return "declining"
    return "flat"


def _level_reliability(
    same_level_results: list[EventingResult],
    one_below_results: list[EventingResult],
    completion_rate: float,
    el_ret_rate: float,
) -> str:
    if len(same_level_results) >= 3 and completion_rate >= 0.75 and el_ret_rate <= 0.2:
        return "proven"
    if len(same_level_results) == 0 and len(one_below_results) >= 2:
        return "stepping up"
    if completion_rate < 0.7 or el_ret_rate > 0.25:
        return "inconsistent"
    return "developing"


def _pattern(results: list[EventingResult], field_name: str) -> str:
    buckets: dict[str, list[float]] = defaultdict(list)
    for result in results:
        if result.is_completed:
            key = getattr(result, field_name) or "Unknown"
            buckets[key].append(result.finishing_score)
    candidates = {key: values for key, values in buckets.items() if len(values) >= 3}
    if not candidates:
        return "not enough data"
    best_key, best_scores = min(candidates.items(), key=lambda item: mean(item[1]))
    return f"best at {best_key} ({round(mean(best_scores), 1)} avg)"


def _same_level(left: str, right: str) -> bool:
    return _level_rank(left) == _level_rank(right) and _level_rank(left) > 0 or _slug(left) == _slug(right)


def _level_rank(level: str) -> int:
    normalized = level.upper().replace("CCI", "CCI")
    match = re.search(r"CCI\s*(\d)", normalized)
    if match:
        return int(match.group(1))
    if "ADVANCED" in normalized:
        return 4
    if "INTERMEDIATE" in normalized:
        return 3
    if "PRELIM" in normalized:
        return 2
    if any(token in normalized for token in ("TRAINING", "NOVICE", "CCN1")):
        return 1
    return 0


def _completion_status(status: str, placing: str = "") -> str:
    text = f"{status} {placing}".lower()
    tokens = set(re.findall(r"[a-z]+", text))
    if tokens & {"el", "elim", "eliminated"}:
        return "eliminated"
    if tokens & {"ret", "retired"}:
        return "retired"
    if tokens & {"wd", "withdrawn"}:
        return "withdrawn"
    return "completed"


def _slug(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")


def _required_str(values: dict[str, object], key: str) -> str:
    value = values.get(key)
    if not isinstance(value, str) or not value:
        raise ValueError(f"{key} must be a non-empty string")
    return value


def _required_number(values: dict[str, object], key: str) -> float:
    value = values.get(key)
    if not isinstance(value, (int, float)):
        raise ValueError(f"{key} must be a number")
    return float(value)


def _optional_str(values: dict[str, object], key: str, default: str = "") -> str:
    value = values.get(key, default)
    if value is None:
        return default
    if not isinstance(value, str):
        raise ValueError(f"{key} must be a string")
    return value


def _optional_number(values: dict[str, object], key: str, default: float) -> float:
    value = values.get(key, default)
    if not isinstance(value, (int, float)):
        raise ValueError(f"{key} must be a number")
    return float(value)


def _nullable_number(values: dict[str, object], key: str) -> float | None:
    value = values.get(key)
    if value is None:
        return None
    if not isinstance(value, (int, float)):
        raise ValueError(f"{key} must be a number or null")
    return float(value)


def _optional_int(values: dict[str, object], key: str, default: int) -> int:
    value = values.get(key, default)
    if not isinstance(value, int):
        raise ValueError(f"{key} must be an integer")
    return value


def _optional_bool(values: dict[str, object], key: str, default: bool) -> bool:
    value = values.get(key, default)
    if not isinstance(value, bool):
        raise ValueError(f"{key} must be a boolean")
    return value

File: src/results.ts
import type { StoredResult } from './scoring';

export const USER_ENTERED_PRIORITY = 100;

export type EventingResultRecord = {
  sourceId: string;
  sourceRecordId: string;
  sourcePriority: number;
  riderName: string;
  horseName: string;
  eventName: string;
  eventDate: string;
  level: string;
  country: string;
  dressageScore: number;
  showJumpingPenalties: number;
  crossCountryJumpPenalties: number;
  crossCountryTimePenalties: number;
  collectedAt: string;
  isUserEntered: boolean;
  riderFeiId?: string;
  horseFeiId?: string;
  combinationId?: string;
  placing?: string;
  showJumpingTimePenalties?: number;
  totalScore?: number | null;
  status?: 'completed' | 'eliminated' | 'retired' | 'withdrawn' | string;
  merStatus?: string;
  eventUrl?: string;
  sourceUrl?: string;
  venue?: string;
};

export type PredictionEvidence = {
  targetLevel: string;
  predictedDressageScore: number;
  predictedXcJumpPenalties: number;
  predictedXcTimePenalties: number;
  predictedShowJumpingPenalties: number;
  predictedShowJumpingTimePenalties: number;
  predictedFinalScoreLow: number;
  predictedFinalScoreHigh: number;
  averageDressageCurrentLevel: number | null;
  averageDressageOneLevelBelow: number | null;
  averageFinalScore: number | null;
  xcClearJumpingRate: number;
  xcTimePenaltyAverage: number;
  sjRailAverage: number;
  completionRate: number;
  eliminationRetirementRate: number;
  bestScore: number | null;
  worstScore: number | null;
  recent3RunAverage: number | null;
  recent5RunAverage: number | null;
  trendDirection: 'improving' | 'flat' | 'declining';
  levelReliability: 'proven' | 'stepping up' | 'inconsistent' | 'developing';
  countryPattern: string;
  venuePattern: string;
  sameLevelResultCount: number;
  horseResultCount: number;
  combinationResultCount: number;
};

export type CombinationPrediction = {
  riderName: string;
  horseName: string;
  likelyFinishingScore: number;
  recentResultCount: number;
  bestRecentScore: number;
  worstRecentScore: number;
  sourceIds: string[];
  confidence: 'low' | 'medium' | 'high';
  predictedLowScore: number;
  predictedHighScore: number;
  evidence: PredictionEvidence;
};

export const SOURCE_LABELS: Record<string, string> = {
  data_fei: 'FEI',
  british_eventing: 'British Eventing',
  equestrian_australia: 'Equestrian Australia',
  equestrian_sports_new_zealand: 'ESNZ',
  usea: 'USEA',
  user_submission: 'My score',
};

export const roundToTenths = (value: number) => Math.round(value * 10) / 10;

export const finishingScore = (result: EventingResultRecord) =>
  result.totalScore ?? roundToTenths(
      result.dressageScore +
        result.showJumpingPenalties +
        (result.showJumpingTimePenalties ?? 0) +
        result.crossCountryJumpPenalties +
        result.crossCountryTimePenalties,
    );

export const slug = (value: string) => value.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');

export const combinationKey = (result: Pick<EventingResultRecord, 'riderName' | 'horseName'>) =>
  `${slug(result.riderName)}::${slug(result.horseName)}`;

export const resultKey = (result: EventingResultRecord) =>
  `${combinationKey(result)}::${slug(result.eventName)}::${result.eventDate}::${slug(result.level)}`;

export const consolidateResults = (results: EventingResultRecord[]) => {
  const selected = new Map<string, EventingResultRecord>();

  for (const result of results) {
    const key = resultKey(result);
    const existing = selected.get(key);

    if (!existing || isBetterResult(result, existing)) {
      selected.set(key, result);
    }
  }

  return [...selected.values()].sort((a, b) => {
    if (a.eventDate !== b.eventDate) {
      return b.eventDate.localeCompare(a.eventDate);
    }

    return `${a.riderName} ${a.horseName}`.localeCompare(`${b.riderName} ${b.horseName}`);
  });
};

export const resultFromStoredResult = (result: StoredResult): EventingResultRecord => ({
  sourceId: 'user_submission',
  sourceRecordId: result.id,
  sourcePriority: USER_ENTERED_PRIORITY,
  riderName: result.rider,
  horseName: result.horse,
  eventName: result.eventName,
  eventDate: result.date,
  level: result.level || 'Unspecified',
  country: result.country || 'N/A',
  dressageScore: result.score.dressagePenalties,
  showJumpingPenalties: result.score.showJumpingPenalties,
  crossCountryJumpPenalties: result.score.crossCountryJumpPenalties,
  crossCountryTimePenalties: result.score.crossCountryTimePenalties,
  collectedAt: result.createdAt,
  isUserEntered: true,
});

export const predictFinishingScore = (
  results: EventingResultRecord[],
  targetCombinationKey: string,
  recentResultLimit = 5,
): CombinationPrediction | null => {
  const recentResults = consolidateResults(results)
    .filter((result) => combinationKey(result) === targetCombinationKey)
    .sort((a, b) => b.eventDate.localeCompare(a.eventDate))
    .slice(0, recentResultLimit);

  if (recentResults.length === 0) {
    return null;
  }
  const evidence = buildPredictionEvidence(results, targetCombinationKey, recentResults[0].level);

  let weightedScoreTotal = 0;
  let weightTotal = 0;
  for (const [index, result] of recentResults.entries()) {
    const weight = recentResults.length - index;
    weightedScoreTotal += finishingScore(result) * weight;
    weightTotal += weight;
  }

  const scores = recentResults.map(finishingScore);
  return {
    riderName: recentResults[0].riderName,
    horseName: recentResults[0].horseName,
    likelyFinishingScore: roundToTenths(weightedScoreTotal / weightTotal),
    recentResultCount: recentResults.length,
    bestRecentScore: Math.min(...scores),
    worstRecentScore: Math.max(...scores),
    sourceIds: [...new Set(recentResults.map((result) => result.sourceId))].sort(),
    confidence: confidence(recentResults.length),
    predictedLowScore: evidence.predictedFinalScoreLow,
    predictedHighScore: evidence.predictedFinalScoreHigh,
    evidence,
  };
};

export const buildPredictionEvidence = (
  results: EventingResultRecord[],
  targetCombinationKey: string,
  targetLevel: string,
): PredictionEvidence => {
  const consolidated = consolidateResults(results);
  const targetResult = consolidated.find((result) => combinationKey(result) === targetCombinationKey);
  const horseSlug = targetResult ? slug(targetResult.horseName) : targetCombinationKey.split('::')[1] ?? '';
  const horseResults = consolidated.filter((result) => slug(result.horseName) === horseSlug);
  const combinationResults = horseResults.filter((result) => combinationKey(result) === targetCombinationKey);
  const evidenceResults = combinationResults.length > 0 ? combinationResults : horseResults;
  const completed = evidenceResults.filter(isCompleted);
  const sameLevel = completed.filter((result) => sameLevelRank(result.level, targetLevel));
  const oneBelow = completed.filter((result) => levelRank(result.level) === levelRank(targetLevel) - 1);
  const scores = completed.map(finishingScore);
  const predictedDressage = weightedPhaseAverage(evidenceResults, targetLevel, (result) => result.dressageScore);
  const predictedXcJump = weightedPhaseAverage(evidenceResults, targetLevel, (result) => result.crossCountryJumpPenalties);
  const predictedXcTime = weightedPhaseAverage(evidenceResults, targetLevel, (result) => result.crossCountryTimePenalties);
  const predictedSjJump = weightedPhaseAverage(evidenceResults, targetLevel, (result) => result.showJumpingPenalties);
  const predictedSjTime = weightedPhaseAverage(evidenceResults, targetLevel, (result) => result.showJumpingTimePenalties ?? 0);
  const predictedTotal = predictedDressage + predictedXcJump + predictedXcTime + predictedSjJump + predictedSjTime;
  const eliminationRetirementRate = rate(
    evidenceResults.filter((result) => ['eliminated', 'retired', 'withdrawn'].includes(completionStatus(result))).length,
    evidenceResults.length,
  );
  const uncertainty = Math.max(2.5, standardDeviation(scores) + eliminationRetirementRate * 8);

  return {
    targetLevel,
    predictedDressageScore: roundToTenths(predictedDressage),
    predictedXcJumpPenalties: roundToTenths(predictedXcJump),
    predictedXcTimePenalties: roundToTenths(predictedXcTime),
    predictedShowJumpingPenalties: roundToTenths(predictedSjJump),
    predictedShowJumpingTimePenalties: roundToTenths(predictedSjTime),
    predictedFinalScoreLow: roundToTenths(Math.max(0, predictedTotal - uncertainty)),
    predictedFinalScoreHigh: roundToTenths(predictedTotal + uncertainty),
    averageDressageCurrentLevel: average(sameLevel.map((result) => result.dressageScore)),
    averageDressageOneLevelBelow: average(oneBelow.map((result) => result.dressageScore)),
    averageFinalScore: average(scores),
    xcClearJumpingRate: rate(completed.filter((result) => result.crossCountryJumpPenalties === 0).length, completed.length),
    xcTimePenaltyAverage: average(completed.map((result) => result.crossCountryTimePenalties)) ?? 0,
    sjRailAverage: roundToTenths(((average(completed.map((result) => result.showJumpingPenalties)) ?? 0) / 4) * 10) / 10,
    completionRate: rate(completed.length, evidenceResults.length),
    eliminationRetirementRate,
    bestScore: scores.length ? Math.min(...scores) : null,
    worstScore: scores.length ? Math.max(...scores) : null,
    recent3RunAverage: average(scores.slice(0, 3)),
    recent5RunAverage: average(scores.slice(0, 5)),
    trendDirection: trend(scores),
    levelReliability: reliability(sameLevel.length, oneBelow.length, rate(completed.length, evidenceResults.length), eliminationRetirementRate),
    countryPattern: pattern(evidenceResults, (result) => result.country),
    venuePattern: pattern(evidenceResults, (result) => result.venue || result.eventName),
    sameLevelResultCount: sameLevel.length,
    horseResultCount: horseResults.length,
    combinationResultCount: combinationResults.length,
  };
};

export const combinationOptions = (results: EventingResultRecord[]) =>
  [...new Map(results.map((result) => [combinationKey(result), result])).entries()]
    .map(([key, result]) => ({
      key,
      label: `${result.horseName} / ${result.riderName}`,
    }))
    .sort((a, b) => a.label.localeCompare(b.label));

export const filterResults = (results: EventingResultRecord[], query: string) => {
  const normalizedQuery = query.trim().toLowerCase();
  if (!normalizedQuery) {
    return results;
  }

  return results.filter((result) =>
    [
      result.horseName,
      result.riderName,
      result.eventName,
      result.level,
      result.country,
      result.status ?? 'completed',
      result.riderFeiId ?? '',
      result.horseFeiId ?? '',
      SOURCE_LABELS[result.sourceId] ?? result.sourceId,
    ]
      .join(' ')
      .toLowerCase()
      .includes(normalizedQuery),
  );
};

export const latestCollectedAt = (results: EventingResultRecord[]) =>
  results.reduce<string | null>((latest, result) => {
    if (!latest || result.collectedAt > latest) {
      return result.collectedAt;
    }

    return latest;
  }, null);

const isBetterResult = (candidate: EventingResultRecord, existing: EventingResultRecord) => {
  if (candidate.sourcePriority !== existing.sourcePriority) {
    return candidate.sourcePriority < existing.sourcePriority;
  }
  if (candidate.isUserEntered !== existing.isUserEntered) {
    return !candidate.isUserEntered;
  }

  return candidate.collectedAt > existing.collectedAt;
};

const confidence = (resultCount: number): CombinationPrediction['confidence'] => {
  if (resultCount >= 5) {
    return 'high';
  }
  if (resultCount >= 3) {
    return 'medium';
  }
  return 'low';
};

const isCompleted = (result: EventingResultRecord) => completionStatus(result) === 'completed';

const completionStatus = (result: EventingResultRecord) => {
  const text = `${result.status ?? ''} ${result.placing ?? ''}`.toLowerCase();
  if (/\b(el|elim|eliminated)\b/.test(text)) {
    return 'eliminated';
  }
  if (/\b(ret|retired)\b/.test(text)) {
    return 'retired';
  }
  if (/\b(wd|withdrawn)\b/.test(text)) {
    return 'withdrawn';
  }
  return 'completed';
};

const weightedPhaseAverage = (
  results: EventingResultRecord[],
  targetLevel: string,
  readValue: (result: EventingResultRecord) => number,
) => {
  const values = results.filter(isCompleted).map((result, index) => {
    const recencyWeight = Math.max(1, results.length - index);
    let levelWeight = sameLevelRank(result.level, targetLevel) ? 1.5 : 0.85;
    if (levelRank(result.level) > levelRank(targetLevel)) {
      levelWeight = 1.2;
    }
    return { value: readValue(result), weight: recencyWeight * levelWeight };
  });
  const weightTotal = values.reduce((total, item) => total + item.weight, 0);
  if (weightTotal === 0) {
    return 0;
  }
  return values.reduce((total, item) => total + item.value * item.weight, 0) / weightTotal;
};

const average = (values: number[]) => {
  if (values.length === 0) {
    return null;
  }
  return roundToTenths(values.reduce((total, value) => total + value, 0) / values.length);
};

const rate = (count: number, total: number) => (total === 0 ? 0 : Math.round((count / total) * 100) / 100);

const standardDeviation = (values: number[]) => {
  if (values.length < 2) {
    return 3;
  }
  const avg = values.reduce((total, value) => total + value, 0) / values.length;
  return Math.sqrt(values.reduce((total, value) => total + (value - avg) ** 2, 0) / values.length);
};

const trend = (scores: number[]): PredictionEvidence['trendDirection'] => {
  if (scores.length < 4) {
    return 'flat';
  }
  const splitAt = Math.max(2, Math.floor(scores.length / 2));
  const newer = average(scores.slice(0, splitAt)) ?? 0;
  const older = average(scores.slice(splitAt)) ?? 0;
  if (newer <= older - 1) {
    return 'improving';
  }
  if (newer >= older + 1) {
    return 'declining';
  }
  return 'flat';
};

const reliability = (
  sameLevelCount: number,
  oneBelowCount: number,
  completionRate: number,
  eliminationRetirementRate: number,
): PredictionEvidence['levelReliability'] => {
  if (sameLevelCount >= 3 && completionRate >= 0.75 && eliminationRetirementRate <= 0.2) {
    return 'proven';
  }
  if (sameLevelCount === 0 && oneBelowCount >= 2) {
    return 'stepping up';
  }
  if (completionRate < 0.7 || eliminationRetirementRate > 0.25) {
    return 'inconsistent';
  }
  return 'developing';
};

const pattern = (results: EventingResultRecord[], readKey: (result: EventingResultRecord) => string) => {
  const buckets = new Map<string, number[]>();
  for (const result of results.filter(isCompleted)) {
    const key = readKey(result) || 'Unknown';
    buckets.set(key, [...(buckets.get(key) ?? []), finishingScore(result)]);
  }
  const candidates = [...buckets.entries()].filter(([, scores]) => scores.length >= 3);
  if (candidates.length === 0) {
    return 'not enough data';
  }
  const [bestKey, bestScores] = candidates.sort((a, b) => (average(a[1]) ?? 0) - (average(b[1]) ?? 0))[0];
  return `best at ${bestKey} (${average(bestScores)?.toFixed(1)} avg)`;
};

const sameLevelRank = (left: string, right: string) =>
  (levelRank(left) > 0 && levelRank(left) === levelRank(right)) || slug(left) === slug(right);

const levelRank = (level: string) => {
  const cci = level.toUpperCase().match(/CCI\s*(\d)/);
  if (cci) {
    return Number(cci[1]);
  }
  const normalized = level.toUpperCase();
  if (normalized.includes('ADVANCED')) {
    return 4;
  }
  if (normalized.includes('INTERMEDIATE')) {
    return 3;
  }
  if (normalized.includes('PRELIM')) {
    return 2;
  }
  if (/(TRAINING|NOVICE|CCN1)/.test(normalized)) {
    return 1;
  }
  return 0;
};
